### PR TITLE
Mk/fix/static analysis

### PR DESCRIPTION
**Pull request by MaggieKimani1**
### Description:
This PR proposes:
- Addition of an EditorConfig file with defined code styling and formatting rules to promote code uniformity across multiple projects
- Configuring SonarCloud in the repo to analyze the codebase and surface any code smells, duplications, bugs and other vulnerabilities.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            "WARNING10": "  * Delete this file",
+            "WARNING11": "  * Open the Visual Studio Code command palette (View->Command Palette)",
+            "WARNING12": "  * run the command: '.NET: Generate Assets for Build and Debug'.",
+            "WARNING13": "",
+            "WARNING14": "If your project requires a more complex launch configuration, you may wish to delete",
+            "WARNING15": "this configuration and pick a different template using the 'Add Configuration...'",
+            "WARNING16": "button at the bottom of this file.",
+            "WARNING17": "*********************************************************************************",
+            "preLaunchTask": "build",
+            "program": "${workspaceFolder}/bin/Debug/<insert-target-framework-here>/<insert-project-name-here>.dll",
```
Comment (baywet) [inline .vscode/launch.json:29]: update this and remove the warning

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      - name: Cache SonarCloud scanner
+        id: cache-sonar-scanner
+        uses: actions/cache@v1
+        with:
+          path: .\.sonar\scanner
+          key: ${{ runner.os }}-sonar-scanner
+          restore-keys: ${{ runner.os }}-sonar-scanner
+      - name: Install SonarCloud scanner
+        if: steps.cache-sonar-scanner.outputs.cache-hit != 'true'
+        shell: powershell
```
Comment (baywet) [inline .github/workflows/build.yml:35]:
```suggestion
        shell: pwsh
```

So it works cross platform

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    types: [opened, synchronize, reopened]
+jobs:
+  build:
+    name: Build
+    runs-on: windows-latest
```
Comment (baywet) [inline .github/workflows/build.yml:11]:
```suggestion
    runs-on: ubuntu-latest
```

performance

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    steps:
+      - name: Set up JDK 11
+        uses: actions/setup-java@v1
```
Comment (baywet) [inline .github/workflows/build.yml:14]:
```suggestion
        uses: actions/setup-java@v2
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        with:
+          java-version: 1.11
```
Comment (baywet) [inline .github/workflows/build.yml:16]:
```suggestion
          distribution: 'adopt'
          java-version: 11
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      - uses: actions/checkout@v2
```
Comment (baywet) [inline .github/workflows/build.yml:17]:
```suggestion
      - uses: actions/checkout@v2.3.4
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        with:
+          fetch-depth: 0  # Shallow clones should be disabled for a better relevancy of analysis
+      - name: Cache SonarCloud packages
+        uses: actions/cache@v1
```
Comment (baywet) [inline .github/workflows/build.yml:21]:
```suggestion
        uses: actions/cache@v2.1.6
```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        with:
+          path: ~\sonar\cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+      - name: Cache SonarCloud scanner
+        id: cache-sonar-scanner
+        uses: actions/cache@v1
```
Comment (baywet) [inline .github/workflows/build.yml:28]:
```suggestion
        uses: actions/cache@v2.1.6
```

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+          New-Item -Path .\.sonar\scanner -ItemType Directory
+          dotnet tool update dotnet-sonarscanner --tool-path .\.sonar\scanner
+      - name: Build and analyze
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        shell: powershell
+        run: |
+          .\.sonar\scanner\dotnet-sonarscanner begin /k:"microsoftgraph_microsoft-graph-devx-api" /o:"microsoftgraph2" /d:sonar.login="${{ secrets.SONAR_TOKEN }}" /d:sonar.host.url="https://sonarcloud.io"
+          dotnet build
```
Comment (baywet) [inline .github/workflows/build.yml:46]:
```suggestion
          dotnet build
          dotnet test MSGraphWebApi.sln --no-build --verbosity normal /p:CollectCoverage=true /p:CoverletOutputFormat=opencover
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 9 above)
Comment (MaggieKimani1) [reply .github/workflows/build.yml:46]: This line helps test the entire solution?

File: .github/workflows/build.yml (hunk 5 above)
Comment (MaggieKimani1) [reply .github/workflows/build.yml:16]: What does this mean?

File: .github/workflows/build.yml (hunk 9 above)
Comment (baywet) [reply .github/workflows/build.yml:46]: runs the unit tests AND exports code coverage to sonarcloud, so sonarcloud can tell you what's covered, what's not and whether a certain pull request is meeting the bar.

Comment (baywet) [reply .github/workflows/build.yml:46]: you still need this line however `./.sonar/scanner/dotnet-sonarscanner end /d:sonar.login="${{ secrets.SONAR_TOKEN }}"`

File: .github/workflows/build.yml (hunk 5 above)
Comment (baywet) [reply .github/workflows/build.yml]: there are multiple distributions of java: the adoptium one, the microsoft one, the oracle one, and many others.

File: .github/workflows/build.yml (hunk 7 above)
Comment (MaggieKimani1) [reply .github/workflows/build.yml]: This change causes an invalid format error during analysis:

image

File: .github/workflows/build.yml (hunk 9 above)
Comment (baywet) [reply .github/workflows/build.yml:46]: slashes are inverted for that last line

File: .github/workflows/build.yml (hunk 7 above)
Comment (baywet) [reply .github/workflows/build.yml]: I'm not sure what happened here but we didn't end up with the suggestion I made, /cache is missing

Comment (MaggieKimani1) [reply .github/workflows/build.yml]: Oh I removed that by mistake when selecting and replacing the version number.

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        if: steps.cache-sonar-scanner.outputs.cache-hit != 'true'
+        shell: pwsh
+        run: |
+          New-Item -Path ./.sonar/scanner -ItemType Directory
+          dotnet tool update dotnet-sonarscanner --tool-path ./.sonar/scanner
+      - name: Build and analyze
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        shell: powershell
```
Comment (baywet) [inline .github/workflows/build.yml:44]:
```suggestion
        shell: pwsh
```

**Review by baywet — APPROVED** [review]

**Pull request merged**